### PR TITLE
Search: Fairer balancing when routing searches by session ID

### LINKED ISSUE #24642: Kibana + 2 data nodes = uneven search loads?
Perhaps a common scenario here where in a 2 data node system one data node gets hammered while the other is idle. The reason seems to lie in default Kibana and search routing config.

* Kibana looks to use a preference key for routing searches based on a session ID. It also uses _msearch for dashboards to bulk up requests.
* All primaries were on one data node and the replicas on the other (not uncommon after a restart)

When elasticsearch gets a request to route on a `preference` setting that is a session ID it looks to select a choice of primary vs replica by [hashing the preference string only](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java#L181). If each shard of an index presents their list of primaries and replicas in the same order (and I haven't confirmed this is the case!) then this routing algo will pick the same node for all searches given the same session key which is what the user was seeing.

If we hashed the preference key AND the shard number we would randomize the choice of primary vs replica and hence node choice for each shard. This would spread load more evenly.

### PR DESCRIPTION
**Pull request by markharwood**
A user reported uneven balancing of load on nodes handling search requests from Kibana which supplies a session ID in a routing preference. Each shardId was selecting the same node for a given session ID from the list of allocations (in their case 2 data nodes with all primaries on one node and replicas on the other).
This change counteracts the tendency to opt for the same node given the same user-supplied preference by incorporating shard ID in the hash of the preference key. This will help randomise node choices across shards.

Closes #24642

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java]: can we use `indexShard.shardId.hashCode()` - it's good to have different hashing across indices as well.

File: core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java]: can we say that the shard ordering is the determined by a fixed node order (and point to the AllocationService for details) and thus we need to make sure we scatter across shards? it's better to say why than just saying better ;)

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: why do we need the cluster service, can't we just use the generated state?

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: can we use random number of replicas, with more than one sometimes ?

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: randomize?

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: can we randomize the session id?

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: can we check that the distribution is balanced in a multi index / multi shard case? like that thing differ at most with 1.

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks Mark. I left some comments

**Review by s1monw — APPROVED** [review]
I agree with @bleskes comments LGTM once they are addressed

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (markharwood) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: Copied over from testPreferNodes(). It looks like it's needed as OperationRouting.getShards(...) fails with Index not found exception without this.

Comment (markharwood) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: Copy and pasted from other preference test.  Will remove

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (markharwood) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: ClusterStateCreationUtils could use an additional helper method to support this?

stateWithActivePrimary(..) lets me pick num replicas but not shards.
stateWithAssignedPrimariesAndOneReplica(...) lets me pick num shards but not replicas

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: sure

Comment (markharwood) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: So with random num shards and replicas I'm going to assume for this balancing test that shard assignments are optimal i.e. num nodes = shards x replicas (or only one Lucene index on each node). Otherwise I can't reason simply about how "bunched up" replicas are on each node and the expected number of nodes servicing a search

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: @markharwood remember to take hash % copies# collisions into account. So if you see two shards being search on the same node, they should have the same hash modulo number of shard copies.

**Comment by markharwood** [comment]
@bleskes I added randomized indexes, shards, sessions etc.
There is now a test for the _minimal desired outcome_ (search spans >1 node) which addresses the original user issue and looks to hold true over thousands of randomized test invocations.
As discussed, a repeatable test for the _optimal outcome_ (searches using all nodes) is not possible due to hash collisions so I implemented the "duel" regression testing approach you outlined where we repeat the hashing logic in the test to ensure there's at least agreement in the routing approach taken.

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: nit: can you use assertThat or expose the actual values in the message.

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: same comment about exposing values.

**Review by bleskes — APPROVED** [review]
It's a shame we had to fall back to a crude duel, but I can't think of anything better.

Thanks @markharwood !

**Comment by markharwood** [comment]
Rebasing before merge

**Comment by ruria** [comment]
I think there are a "few" 2 data nodes clusters out there using Kibana. All we are using 50% of resources!! Rebalancing primary shards is the only way to go (I think) and it seems to me like a CRITICAL issue, maybe a backport is worth considering.

We ´ve tried to rebalance prymary shards acrros cluster using cluster.routing.rebalance.enable but it doesn´t work for us. I´m little confused here ( https://github.com/elastic/elasticsearch/issues/3293 ) Is it possible to rebalnace primary shards?

**Comment by ruria** [comment]
Upss! Rebalancing primary shards not solves the issue, forget this. A user using Kibana (2 data node) is using just 50% of resources, and it have not workarround.

**Comment by bleskes** [comment]
@ruria no, rebalancing the primaries wouldn't work. Do you only have one kibana users? multiple sessions should spread out the load.

**Comment by ruria** [comment]
Yes, I have a few users. But for each of them, the "power" used is the half. Not all users are searching concurrently.

In our scenario, migrating to new hardware (x3 faster at indexing) from 2.3 to 5.4, we´re suffering a huge search time slowdown.

**Pull request merged**
